### ISSUE TITLE
Why not a Helix Mode, to go along with Vi and Emacs?

### ISSUE THREAD

**Issue by luccahuguet** [nushell/reedline#639]
I think a helix-mode would go a long way into making the reedline experience more complete, now targeting the three major modal editors.

Helix's philosophy is a bit like this (a small excerpt from their [vision page](https://github.com/helix-editor/helix/blob/master/docs/vision.md) on github):

> Selection -> Action, not Verb -> Object. Interaction models aren't linguistics, and "selection first" lets you see what you're doing (among other benefits).
> We aren't playing code golf. It's more important for the keymap to be consistent and easy to memorize than it is to save a key stroke or two when editing.

helix also has multi-cursors but we could add it later. 

Also important to note that helix is heavily inspired by kakoune, so this would be satisfying for that group of users too.

I added a comment in the keybinding discussion: https://github.com/nushell/reedline/issues/69#issuecomment-1736309196

and if needed a could open an issue in nushell's main repo. 

EDIT: 
One can easily set the EDITOR environment variable to hx and then press `ctrl o` while typing on nushell
This will open a temporary buffer in helix, and there you go, helix mode
So this diminishes the case of this feature, but it still exists, of course!

**Comment by fdncred** [comment]
Helix mode would be fun. All it takes is someone to come in and contribute it.

**Comment by luccahuguet** [comment]
@fdncred Terrific!

I hope someone picks this up

**Comment by alxpettit** [comment]
I just wanted to say I would LOVE a Helix editing mode! I actually already find myself unconsciously trying to use Helix keybindings to edit the prompt line in Nushell when I'm tired. :P

**Comment by MatrixManAtYrService** [comment]
I want to include a helix mode in an app that I'm working on.  When I get my head all the way around it, I'll try to add it here too, unless somebody has beat me to it.

That said, my Rust is not yet strong, so it'll be quite a long time before I'm ready for that contribution.  🤞 somebody gets to it first.

**Comment by MatrixManAtYrService** [comment]
I saw a comment in [another issue](https://github.com/nushell/nushell/issues/2878#issue-780778831):

> The Vi mode support seems to rely on rustyline which unfortunately doesn't have support for visual mode https://github.com/kkawakam/rustyline/issues/335

Is that still true?  Helix editing relies pretty heavily on the ability to do visual-mode like stuff (each movement is also a selection), so I'm worried that that limitation would also limit helix mode.

**Comment by luccahuguet** [comment]
@sholderbach added some helpful comments in the draft by @savente93 on what is probably needed before a PR concerning this issue can be done: https://github.com/nushell/reedline/pull/642#issuecomment-1742705559

Just putting this here to make it easier to track. It's been a while, maybe there's been progress in some of the items..

> * We have maybe reduced the number of `EditCommand` implementations that do essentially the same in slight variations: `Move`/`Cut`/`Delete`:



> I think what I would gather as a reasonable point where it makes sense to go ahead with helix mode without impacting the effort to getting the rest of the modes nice would be:
> 
> * We have maybe reduced the number of `EditCommand` implementations that do essentially the same in slight variations: `Move`/`Cut`/`Delete`: https://github.com/nushell/reedline/blob/b1344f6a653b7c37154a1dd79b3c4e34afe6cfc9/src/core_editor/editor.rs#L43-L98
> * We have a selection implementation in at least one of the existing modes (be it visual mode for vim or Shift-Arrow for the non modal emacs-like mode)
> * We are happy with how you configure movements that can affect both the buffer and the general UI (e.g. up/down for history, completion or other menus) We currently have this `UntilFound` logic that has some implicit rules in how you need to order things and has some edge cases (multiplier motions in vi, left and right movements while in the history scrollback). This is not super critical for the helix implementation concerns itself but I think it makes sense that we have some clarity in how things should be configured on the nushell side so we can drop helix mode in when it is ready (potentially after we stabilize a config format for 1.0)

**Comment by luccahuguet** [comment]
I will also leave this here: a vscode extension that adds some helix commands. This could bring some insight 

https://github.com/jasonwilliams/vscode-helix

**Comment by haxfn** [comment]
Interesting Issue.

**Comment by luccahuguet** [comment]
Added an edit to the issue text:

One can easily set the EDITOR environment variable to hx and then press ctrl o while typing on nushell
This will open a temporary buffer in helix, and there you go, helix mode
So this diminishes the case of this feature, but it still exists, of course!

**Comment by eugenesvk** [comment]
Is there no way to actually use (Neo)Vim and Helix and ... for the editing instead of trying to reimplement a subset of their commands? The difference from the Ctrl-O suggestion would be to have those editors have no separate UI, they'd just  operate within that one single input line "behind the scenes".

This would, for example, give you all the syntax-specific-text-object rich movements so that in multiline mode you could have a command to move to the beginning of a nushell function, not just move up/down/by words.

**Comment by philocalyst** [comment]
There was a PR recently merged for this, #1039, which means that this issue should be closed?

**Comment by fdncred** [comment]
Helix mode is not complete yet.

**Comment by luccahuguet** [comment]
> There was a PR recently merged for this, [#1039](https://github.com/nushell/reedline/pull/1039), which means that this issue should be closed?

awesome! thanks for the info

**Comment by eaves-dropper** [comment]
Helix mode isn't available on latest release, even though it was merged on Apr 17

**Comment by kronberger-droid** [comment]
Helix mode is still WIP right now and behind a feature flag.
Since Helix's semantics are fundamentally different to Vi or Emacs we have to rethink many assumptions we made in the Editor and get some things in order before we can cleanly integrate it.

So yeah helix mode is not available as of now and will still take some time, but I am as excited as you are for it.
We will let you all know when it is in a state where you can start testing and give us feedback.

**Comment by luccahuguet** [comment]
the nushell team rocks. helix mode will be awesome. thanks

**Comment by luccahuguet** [comment]
Possibly closed by https://github.com/nushell/reedline/pull/1138

**Comment by fdncred** [comment]
closed by https://github.com/nushell/reedline/pull/1138
